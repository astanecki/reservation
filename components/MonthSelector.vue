<template>
  <div class="month-selector">
    <span
      class="month-selector__arrow"
      @click="changeMonth(-1)"
    >
      &#8249;
    </span>
    <span class="month-selector__header">
      {{ headerLabel }}
    </span>
    <span
      class="month-selector__arrow"
      @click="changeMonth(+1)"
    >
      &#8250;
    </span>
  </div>
</template>
<script>
import moment from 'moment';

import CONFIG from '~/config/config.json';

export default {
  name: 'MonthSelector',

  props: {
    value: {
      type: Object,
      required: true,
    },
  },

  computed: {
    headerLabel() {
      const { YEAR, MONTH } = CONFIG.DATE_FORMATS;

      return `${this.value.format(MONTH)} ${this.value.format(YEAR)}`;
    },
  },

  methods: {
    /**
     * @function
     * @param {Number} direction
     */
    changeMonth(direction) {
      this.$emit('input', moment(this.value).add(direction, 'month'));
    },
  }
}
</script>
<style lang="scss" scoped>
.month-selector {
  display: flex;
  justify-content: space-between;
  align-items: center;

  color: white;
  font-weight: 700;
  background-color: #17d09e;
  padding: 10px;

  &__arrow {
    cursor: pointer;

    padding: 0 5px;
    color: black;
    font-size: 28px;
  }
}
</style>
