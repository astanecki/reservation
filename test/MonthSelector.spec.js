import { mount } from "@vue/test-utils";
import { advanceTo, clear } from "jest-date-mock";
import moment from "moment";

import MonthSelector from "~/components/MonthSelector";

describe("MonthSelector", () => {
  const requiredProps = {};
  const createWrapper = (methods = {}) => {
    return mount(MonthSelector, {
      propsData: requiredProps,
      methods
    });
  };

  beforeEach(() => {
    advanceTo("2000-01-01");

    requiredProps.value = moment();
  });

  afterEach(() => {
    clear();
  });

  it("is a Vue instance", () => {
    const wrapper = createWrapper();

    expect(wrapper.vm).toBeTruthy();
  });

  it(`should match initial snapshot`, () => {
    const wrapper = createWrapper();

    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  describe("computed", () => {
    let wrapper;

    beforeEach(() => {
      wrapper = createWrapper();
    });

    it(`should return proper headerLabel value`, () => {
      expect(wrapper.vm.headerLabel).toMatchInlineSnapshot(`"January 2000"`);
    });
  });

  describe("methods", () => {
    let wrapper;

    describe("changeMonth", () => {
      it(`should call $emit`, () => {
        wrapper = createWrapper();

        wrapper.vm.$emit = jest.fn();

        wrapper.vm.changeMonth(+1);

        expect(wrapper.vm.$emit).toHaveBeenCalledWith(
          "input",
          moment(requiredProps.month).add(1, "month")
        );
      });

      it(`should be called when first arrow fired click event`, () => {
        const selfSpy = jest.fn();

        wrapper = createWrapper({
          changeMonth: selfSpy
        });

        wrapper
          .findAll(".month-selector__arrow")
          .at(0)
          .trigger("click");

        expect(selfSpy).toHaveBeenCalledWith(-1);
      });

      it(`should be called when second arrow fired click event`, () => {
        const selfSpy = jest.fn();

        wrapper = createWrapper({
          changeMonth: selfSpy
        });

        wrapper
          .findAll(".month-selector__arrow")
          .at(1)
          .trigger("click");

        expect(selfSpy).toHaveBeenCalledWith(+1);
      });
    });
  });
});
