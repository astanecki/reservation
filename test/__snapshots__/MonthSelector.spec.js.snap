// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthSelector should match initial snapshot 1`] = `
<div
  class="month-selector"
>
  <span
    class="month-selector__arrow"
  >
    
    ‹
  
  </span>
   
  <span
    class="month-selector__header"
  >
    
    January 2000
  
  </span>
   
  <span
    class="month-selector__arrow"
  >
    
    ›
  
  </span>
</div>
`;
